#! /bin/sh /usr/share/dpatch/dpatch-run

## 06_nissl_dvbspu.dpatch by Reinhard Nissl <rnissl@gmx.de>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: SPU fixes (vdr-dvd)

@DPATCH@
diff -urNad --exclude=CVS --exclude=.svn ./dvbspu.c /tmp/dpep-work.B5kxnx/vdr-1.3.28/dvbspu.c
--- ./dvbspu.c	2005-08-07 14:06:32.000000000 +0200
+++ /tmp/dpep-work.B5kxnx/vdr-1.3.28/dvbspu.c	2005-08-07 23:20:39.000000000 +0200
@@ -336,6 +336,20 @@
     return size;
 }
 
+static bool OsdMatchesArea(cOsd *osd, tArea &area)
+{
+    cBitmap *bmp = osd->GetBitmap(0);
+    if (!bmp)
+       return false;
+    if (bmp->Bpp() < area.bpp)
+       return false;
+    if (bmp->X0() > area.x1 || bmp->Y0() > area.y1)
+       return false;
+    if ((bmp->X0() + bmp->Width()) <= area.x2 || (bmp->Y0() + bmp->Height()) <= area.y2)
+       return false;
+    return true;
+}
+
 void cDvbSpuDecoder::Draw(void)
 {
     if (!spubmp) {
@@ -366,12 +380,16 @@
        }
 
     if (bg || fg) {
+        int x2 = areaSize.x2;
+        while ((x2 - areaSize.x1 + 1) & 0x03)
+              x2++;
+        tArea Area = { areaSize.x1, areaSize.y1, x2, areaSize.y2, (fg && bg) ? 4 : 2 };
+        if (osd && !OsdMatchesArea(osd, Area)) {
+           delete osd;
+           osd = NULL;
+           }
         if (osd == NULL) {
            osd = cOsdProvider::NewOsd(0, 0);
-           int x2 = areaSize.x2;
-           while ((x2 - areaSize.x1 + 1) & 0x03)
-                 x2++;
-           tArea Area = { areaSize.x1, areaSize.y1, x2, areaSize.y2, (fg && bg) ? 4 : 2 };
            osd->SetAreas(&Area, 1);
            }
 
